torch/_inductor/fx_passes/post_grad.py

```diff
@@ -802,3 +805,162 @@ def repl(inp, x1, x2):
+        return True
+
+    if not isinstance(
+        node.target, torch._ops.OpOverload
+    ) and node.target.namespace not in ("prims", "aten"):
+        return True
+
+    # can take mixed devices - bail for now
+    incompatible_ops = (aten._embedding_bag.default,)
+
+    return node.target in incompatible_ops
+
+
+def get_node_device(node: fx.Node) -> Optional[torch.device]:
+    ten = node.meta.get("val")
+    return None if not isinstance(ten, torch.Tensor) else ten.device
+
+
+def get_cpu_indeg_count(graph) -> Dict[fx.Node, int]:
+    cpu_indeg: Dict[fx.Node, int] = Counter()
+
+    for node in graph.nodes:
+        cpu_count = 0
+
+        def add_cpu_inp(node):
+            nonlocal cpu_count
+            device = get_node_device(node)
+            cpu_count += device is not None and device.type == "cpu"
+
+        pytree.tree_map_only(torch.fx.Node, add_cpu_inp, (node.args, node.kwargs))
+
+        if cpu_count:
+            cpu_indeg[node] = cpu_count
+
+    return cpu_indeg
+
+
+def move_constructors_to_cuda(graph):
+    """
+    Moves intermediary tensors which are constructed on the cpu to cuda when safe
+    """
+    if not torch.backends.cuda.is_built():
+        return
+
+    cuda_devices = set()
+    constructors = []
+
+    for node in graph.nodes:
+        device = get_node_device(node)
+        if device and device.type == "cuda":
+            cuda_devices.add(device)
+
+        if not isinstance(
+            node.target, torch._ops.OpOverload
+        ) or node.target.namespace not in ("prims", "aten"):
+            continue
+
+        if not torch._subclasses.fake_tensor._is_tensor_constructor(node.target):
+            continue
+
+        if not node.kwargs.get("device") == torch.device("cpu"):
+            continue
+
+        constructors.append(node)
+
+    # not handling multiple cuda devices initially
+    if not constructors or len(cuda_devices) != 1:
+        return
+
+    movable_constructors = find_movable_constructors(graph, constructors)
+
+    for node in movable_constructors:
+        kwargs = node.kwargs.copy()
+        kwargs["device"] = next(iter(cuda_devices))
+        node.kwargs = kwargs
+
+
+def find_movable_constructors(graph, constructors: List[fx.Node]) -> Set[fx.Node]:
+    """
+    Starting from the cpu constructors, iterate through the graph and test that all of their
+    downstream uses can safely be moved to cpu.
+    """
+    cpu_indeg: Dict[fx.Node, int] = get_cpu_indeg_count(graph)
+
+    # which constructors cannot be moved to cuda
+    cannot_move_to_cuda: Set[fx.Node] = set()
+
+    # For any node in the graph, which constructors does it have a dependency on
+    constructor_dependencies: Dict[fx.Node, Set[fx.Node]] = defaultdict(set)
+
+    # if a cpu node has a dependency on two different cpu constructors,
+    # then if either constructor cannot be moved to cuda, the other cannot as well.
+    # In this case any node with a dependency on one will have a dependency on the other
```

Comment:
Unless you're willing to duplicate constructor calls, no?

Reply:
While that is true, we're mostly optimizing for cudagraphs so going to leave this unimplemented for now